Scroll To Text Fragment Indicator flashes though banners.

megangardner opened the pull request:
#### 13be919f4512dc3771840c699abf5389bbe422e9
<pre>
Scroll To Text Fragment Indicator flashes though banners.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245116">https://bugs.webkit.org/show_bug.cgi?id=245116</a>
rdar://99669383

Reviewed by Tim Horton.

The text indicator is flashing through banners that are over the text at the time of
flash. In order to stop this, check the edge of the indicator and do not start it if
either end is occluded. This does not deal with popups that show up part way through
the flash of the indicator.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::textFragmentIndicatorTimerFired):

Canonical link: <a href="https://commits.webkit.org/254494@main">https://commits.webkit.org/254494@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8539862b8096b52b6f7b17c7e87413f20956eb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89222 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33783 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98536 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154854 "Hash d8539862 for PR 4291 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93230 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32289 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/27829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81581 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93002 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94869 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25642 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76137 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25578 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80509 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68553 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30066 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29791 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3154 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33238 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38461 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31941 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/dom/SimpleRange.cpp:
```diff
@@ -301,7 +301,7 @@ template<TreeType treeType> bool intersects(const SimpleRange& range, const Node
     if (!nodeRange)
         return contains<treeType>(node, range.start.container);
-    return is_lt(treeOrder<treeType>(nodeRange->start, range.end)) && is_lt(treeOrder<treeType>(range.start, nodeRange->end));
+    return intersects(range, nodeRange.value());
```
probably `*nodeRange` is more normal, but this is fine

hortont424 replied:
This should be explained in the changelog.

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2449,6 +2449,7 @@ void FrameView::scrollToFocusedElementInternal()
     FrameView::scrollRectToVisible(absoluteBounds, *renderer, insideFixed, { m_selectionRevealModeForFocusedElement, ScrollAlignment::alignCenterIfNeeded, ScrollAlignment::alignCenterIfNeeded, ShouldAllowCrossOriginScrolling::No });
 }
 
+
```
Please remove this extraneous newline

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2466,9 +2467,40 @@ void FrameView::textFragmentIndicatorTimerFired()
+        }
+        document.page()->chrome().client().setTextIndicator(textIndicator->data());
+    }
     cancelScheduledTextFragmentIndicatorTimer();
```
Move this up before the if, invert the if, and make it an early return?

megangardner replied:
I didn't do an early return because we do not want to cancel the flash if we have only searched for a character.

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2466,9 +2467,40 @@ void FrameView::textFragmentIndicatorTimerFired()
+        return;
+    
+    if (textIndicator) {
+        constexpr OptionSet<HitTestRequest::Type> hitType { HitTestRequest::Type::ReadOnly, HitTestRequest::Type::Active, HitTestRequest::Type::AllowVisibleChildFrameContentOnly };
```
I'm not sure why `AllowVisibleChildFrameContentOnly`. Also, do you want one to allow you to descend into subframes and/or shadow dom?

megangardner replied:
I wanted to be generous with what could block the content, and I thought this was a way to be generous.

hortont424 replied:
Oh, I was suggesting being *more* generous

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2466,9 +2467,40 @@ void FrameView::textFragmentIndicatorTimerFired()
+        auto firstCharRange = resolveCharacterRange(range, { 0, 1 });
+        auto firstCharRect = RenderFlexibleBox::absoluteTextRects(firstCharRange);
+        
+        auto collapsedStart = SimpleRange(BoundaryPoint(range.startContainer(), range.startOffset()), BoundaryPoint(range.startContainer(), range.startOffset()));
```
Paragraphing here is ... confusing

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2466,9 +2467,40 @@ void FrameView::textFragmentIndicatorTimerFired()
+        if (!intersects(range, *result.targetNode()))
+            return;
+
+        if (range.endOffset() > 1) {
```
Odd that we can't test the end if it ends just inside a node (can that happen?).

megangardner replied:
I think if we look for a single character word this can happen (don't know why you would, but it is valid)

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2466,9 +2467,40 @@ void FrameView::textFragmentIndicatorTimerFired()
+
+            result = page->mainFrame().eventHandler().hitTestResultAtPoint(LayoutPoint(lastCharRect.first().center()), hitType);
+            if (!intersects(range, *result.targetNode()))
+                return;
```
Do you mean to bail entirely? (should cancelScheduledTextFragmentIndicatorTimer still happen?)

megangardner replied:
That's probably right, I'll need to break or something.

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2457,20 +2457,49 @@ void FrameView::textFragmentIndicatorTimerFired()
+            return;
+
+        if (range.endOffset() > 1) {
+            auto lastCharRange = resolveCharacterRange(range, { range.endOffset() - 1, 1 });
```
I am pretty sure this is wrong, since endOffset is in the endContainer's space, and resolveCharacterRange's CharacterRange is in the space of all characters in the range.

hortont424 replied:
Maybe rniwa or darin know the "right" way to do this!

hortont424 replied:
(and by "wrong", I mean you are picking a totally random end character depending on the structure of the DOM covered by the range)

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2457,20 +2457,54 @@ void FrameView::textFragmentIndicatorTimerFired()
+        if (!intersects(range, *result.targetNode()))
+            return;
+        
+        if (textRects.size() > 2) {
```
I think this should be >=, right?

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2457,20 +2457,54 @@ void FrameView::textFragmentIndicatorTimerFired()
+                return;
+        }
+        
+        if (textRects.size() > 4) {
```
I think this should be >= also?

megangardner committed:
Scroll To Text Fragment Indicator flashes though banners.
https://bugs.webkit.org/show_bug.cgi?id=245116
rdar://99669383

Reviewed by Tim Horton.

The text indicator is flashing through banners that are over the text at the time of
flash. In order to stop this, check the edge of the indicator and do not start it if
either end is occluded. This does not deal with popups that show up part way through
the flash of the indicator.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::textFragmentIndicatorTimerFired):

Canonical link: https://commits.webkit.org/254494@main
